### DISCUSSION TITLE
Docker images and AMIs

### DISCUSSION THREAD

**Discussion by hazmeister** [RedashCommunity/redash#2] [General]
AMIs and other cloud images are based on Ubuntu 18 which is almost EOL. I looked at updating them, but essentially it's a image with docker compose and a couple of convenient shell scripts which I think could be documented instead and hosted in a reference repo.

I could write a terraform module for hosting with ECS and RDS since we sort of recommend using that anyway once you reach a certain size of deployment.

We're going to need hosting for docker images- a single account is free on docker hub if we can work with a single set of credentials, until we can join the [Docker Open Source](https://www.docker.com/community/open-source/application/) program.

I think GitHub and quay.io also have free tiers.

**Comment by justinclift** [comment]
Cool, good thinking. :smile:

The EOL for Ubuntu 18 seems to be literally this month (!):

https://ubuntu.com/blog/ubuntu-18-04-eol-for-devices

**Probably** the easier update we can do in the short term is move the image to Ubuntu 20 (LTS).  Should be less changes than a newer Ubuntu.

The "document it" thing sounds like a good idea too.

Do you want to keep on poking around options for the hosting side of things, and set up our Docker credentials?

**Reply by hazmeister** [reply]
Sure thing - I've parked the redashcommunity username on docker hub.

**Reply by justinclift** [reply]
Awesome. :smile:

**Reply by hazmeister** [reply]
Started off by pushing a [one off build of version 10](https://hub.docker.com/r/redashcommunity/redash/tags) with the pysaml2 fix as per the instructions [here](https://github.com/getredash/redash/security/advisories/GHSA-rm5x-rgmf-qv5c).  I'll write a proper README tomorrow and start looking at getting CI to push.

**Reply by justinclift** [reply]
Uh oh.  *Please* increment that version number ASAP, so it doesn't directly conflict with the existing version 10.1 of the original Redash repo!

Maybe make it 10.2.0 or something?  i.e. Something so less savvy users can feel confident they picked the right thing.

Am thinking that using the same version number as the existing Redash release will cause problems and confusion otherwise. :wink:

**Reply by justinclift** [reply]
As a thought, the version numbers that show up in the Redash user interface are stored in [`package.json`](https://github.com/RedashCommunity/redash/blob/a5d12b1740402a25dce6ded436af5e3fbd227d79/package.json#L3) and [`redash/__init__.py`](https://github.com/RedashCommunity/redash/blob/a5d12b1740402a25dce6ded436af5e3fbd227d79/redash/__init__.py#L18).

I'm not sure if there's some process that automatically updates them prior to a release, but we can always update them manually before building the new docker image as we develop our own release processes. :smile:

**Reply by hazmeister** [reply]
OK, pulled them down for the minute whilst we figure this out.  I'll look at updating version numbers, although to migrate to our images, people will be changing tag anyway, so it's a deliberate change.  If I can patch the images for v8 and v9 , should they be published as 8.1 and 9.1?

**Reply by justinclift** [reply]
Oh, that's a good point. :smile:

I hadn't personally thought about the older releases, but it makes good sense to do them as well where possible.

For "what exact version numbers though"... good question.

Just pulling numbers out of thin air, maybe lets go with x.5 to indicate "some kind of big change, but not enough for a whole release number"?

So:

Old Redash | Updated Redash Community build
-- | --
10.1 | 10.5
9.0 | 9.5
8.0 | 8.5

Thoughts?

**Reply by justinclift** [reply]
@hazmeister Potential beta testing volunteer: @joshuacronemeyer (from [here](https://github.com/RedashCommunity/redash/issues/8#issuecomment-1544723738)).

**Reply by hazmeister** [reply]
Can do that. Looks like we ought to get some ami upgrade instructions written too. As the image is basically git source and docker compose it should be easy enough.

**Comment by rpmcoding** [comment]
This could do with updating also - https://marketplace.digitalocean.com/apps/redash

Not sure if we have access though.

**Reply by hazmeister** [reply]
Don't think we can update those but we might be able to setup new ones.  They're all built from the same source.

**Reply by justinclift** [reply]
Oh wow.  Version 8.

Once we get our initial release sorted out, we can reach out to Digital Ocean about that.

I have some contacts with Digital Ocean's "Open Source" program team - mainly used for sponsorships and such - and they might be able to assist.  Not guaranteed, but it doesn't hurt to try. :smile:

**Reply by hazmeister** [reply]
On the subject of v8, 9 and 10 we're never considered stable (or at least that was the impression I got). I know SAML is broken in later versions (not just out of date) but are there any other major issues? Or should we be focusing on getting folks to migrate to v10?

**Reply by justinclift** [reply]
Hmmm, forcing an update to v10 for places that haven't already done it, seems like it'd be a dick move on our part.  I'd kind of expect that any place not yet running v10 is likely doing that for a reason.

That being said, we can't be reasonably expected to support every old release forever.

So my first thought would be for us to get these initial security-focused releases out the door first. eg:

* Version 10 security release
* Versions 8 & 9 security releases <-- only if they can be done using reasonable effort

That'd give us breathing space to think about the pros and cons of other things for future stuff. :smile:

**Reply by justinclift** [reply]
@hazmeister On the topic of building these security-fixing releases, you're doing so from the existing `release/*` branches yeah?  eg `release/10.x`, `release/9.x`, etc

Asking because I haven't looked into the patches in those branches yet.  No idea at all what state they're in (etc). :innocent:

**Reply by hazmeister** [reply]
Originally I was monkey patching- grabbing the original images and adding some fixes in new layers. If we have any other planned fixes in the pipeline it would make sense to build from source.

**Reply by justinclift** [reply]
If we can build from source, it's *probably* the way to go... though I'm somewhat suspecting the older versions of dependencies are going to create their own kind of hassles.

Oh well, if you hit issues with stuff then er... open Issues here?  There are people around who are willing to help out, especially to get an updated working release out the door. :smile:

**Reply by joshuacronemeyer** [reply]
> Hmmm, forcing an update to v10 for places that haven't already done it, seems like it'd be a dick move on our part. I'd kind of expect that any place not yet running v10 is likely doing that for a reason.

Just FYI the only reason we're still on 8 is that the instructions on the website seemed to encourage using the AMI, which is on 8. IMO as long as there are decent instructions for moving to 10 I don't mind upgrading. Also, I don't require an AMI. My goal is to stay on a nice little documented path to maintaining our redash instance.

**Reply by justinclift** [reply]
Cool, that makes good sense. :smile:

**Comment by tinder-tder** [comment]
I have a terraform module we can probably share (getting clearance) but its basically AWS based, ec2 (ubuntu+docker-compose) + rds + elasticache + alb + kms.  We can adapt it  for ecs too if needed.  ubuntu 22 lts  works fine

**Reply by justinclift** [reply]
That sounds potentially useful. :smile:

Is the concept that we use it for building our images, plus also publish it for use by the Redash Community?

**Reply by tinder-tder** [reply]
I was talking about a reference AWS deployment, if we want to build images i have a terraform+packer module we can repurpose to build the redash ami/container.  but we can use it to push to ecr or whatever.  We have a pretty big IaC library but we are very AWS centric.  so it may not be good for the general public

**Reply by justinclift** [reply]
Ahhhh, got it.  That makes sense.

Yeah, that could be useful down the track a bit. :smile:

**Reply by tinder-tder** [reply]
I had to do a bunch of cleanup/fixes on the setup repo to get things running the first time we deployed.  So when I get time I will try and cleanup and update that repo if we want to keep it there.  I am not sure the prev repo org was necessary and long term maintainable (ie one repo referencing another with stale links/info)

**Reply by justinclift** [reply]
Yeah, from rough memory I think the separate setup repo kind of grew out from its original home of being a "setup" directory in the main Redash repo.

But then it was kind of left to rot, which is likely the main problem. :wink:

**Reply by tinder-tder** [reply]
Yea, I think I would prefer to keep it in the main repo but we should come up with a consensus.  I don't want to rock the boat cause this isn't exactly greenfield.

**Reply by justinclift** [reply]
Yeah, I don't have a strong opinion either way, so "whatever works" from my point of view. :smile:

**Comment by Jengro777** [comment]
https://hub.docker.com/r/redashcommunity/redash
Looking forward to it working as soon as possible

**Reply by Jengro777** [reply]
That's awesome. It's already having an impact.

**Comment by gaecoli** [comment]
@justinclift I build a pre version based on master code. But i don't have permission to push into our docker hub.

**Reply by justinclift** [reply]
@gaecoli Cool.  I think @hazmeister was a bit faster, as he has added an automatic task to our GitHub repo now: https://github.com/RedashCommunity/redash/commit/3f586da1da8451f6c526c47bf0eb4817ddc8dd0e

This should automatically push to our Docker Hub repo from now on.

That being said, the username/password needed should be in our repo secrets somewhere, so you should be able to get them and do stuff. :smile:

**Reply by gaecoli** [reply]
> @gaecoli Cool. I think @hazmeister was a bit faster, as he has added an automatic task to our GitHub repo now. It should automatically push to our Docker Hub repo from now on.
> 
> That being said, the username/password needed should be in our repo secrets somewhere, so you should be able to get them and do stuff. 😄

ok, maybe docker hub username and password should be publiced to developers org.

**Reply by justinclift** [reply]
Good point.  Are you ok to do that in the existing private area we have?

**Reply by gaecoli** [reply]
> Good point. Are you ok to do that in the existing private area we have?

yes
